client/app/services/query-result.js

```diff
@@ -35,418 +39,398 @@ function getColumnFriendlyName(column) {
   return getColumnNameWithoutType(column).replace(/(?:^|\s)\S/g, a => a.toUpperCase());
 }
 
-function QueryResultService($resource, $timeout, $q, QueryResultError, Auth) {
-  const QueryResultResource = $resource("api/query_results/:id", { id: "@id" }, { post: { method: "POST" } });
-  const QueryResultByQueryIdResource = $resource("api/queries/:queryId/results/:id.json", {
-    queryId: "@queryId",
-    id: "@id",
-  });
-  const Job = $resource("api/jobs/:id", { id: "@id" });
-  const JobWithApiKey = $resource("api/queries/:queryId/jobs/:id", { queryId: "@queryId", id: "@id" });
-  const statuses = {
-    1: "waiting",
-    2: "processing",
-    3: "done",
-    4: "failed",
-  };
-
-  function handleErrorResponse(queryResult, response) {
-    if (response.status === 403) {
-      queryResult.update(response.data);
-    } else if (response.status === 400 && "job" in response.data) {
-      queryResult.update(response.data);
-    } else {
-      logger("Unknown error", response);
-      queryResult.update({
-        job: {
-          error: response.data.message || "unknown error occurred. Please try again later.",
-          status: 4,
-        },
-      });
-    }
+const createOrSaveUrl = data => (data.id ? `api/query_results/${data.id}` : "api/query_results");
+const QueryResultResource = {
+  get: ({ id }) => axios.get(`api/query_results/${id}`),
+  post: data => axios.post(createOrSaveUrl(data), data),
+};
+
+const statuses = {
+  1: "waiting",
+  2: "processing",
+  3: "done",
+  4: "failed",
+};
+
+function handleErrorResponse(queryResult, response) {
+  if (response.status === 403) {
+    queryResult.update(response.data);
+  } else if (response.status === 400 && "job" in response.data) {
+    queryResult.update(response.data);
+  } else {
+    logger("Unknown error", response);
+    queryResult.update({
+      job: {
+        error: response.data.message || "unknown error occurred. Please try again later.",
+        status: 4,
+      },
+    });
   }
+}
 
-  class QueryResult {
-    constructor(props) {
-      this.deferred = $q.defer();
-      this.job = {};
-      this.query_result = {};
-      this.status = "waiting";
+class QueryResult {
+  constructor(props) {
+    this.deferred = $q.defer(); // TODO: replace this defer
```

Comment:
`$q.defer()` is and old API that was later replaced with `Promise`. Replacement pattern is something like:
```js
function defer() {
  const result = {
    promise: new Promise((resolve, reject) => {
      result.resolve = resolve;
      result.reject = reject;
    });
  };
  return result;
}
```

Reply:
I was actually thinking if I was going to keep the defer pattern with something like that or replace it with something else